shanoir-ng-datasets/src/main/java/org/shanoir/ng/dataset/service/DatasetServiceImpl.java

```diff
@@ -115,33 +115,72 @@ public class DatasetServiceImpl implements DatasetService {
 
 	private static final Logger LOG = LoggerFactory.getLogger(DatasetServiceImpl.class);
 
+	private void delete(Dataset entity) throws ShanoirException, SolrServerException, IOException, RestServiceException {
+		Long id = entity.getId();
+
+		// Remove parent processing to avoid errors
+		entity.setDatasetProcessing(null);
+		processingService.removeDatasetFromAllProcessingInput(id);
+		processingResourceService.deleteByDatasetId(id);
+		propertyService.deleteByDatasetId(id);
+		repository.deleteById(id);
+
+		shanoirEventService.publishEvent(new ShanoirEvent(ShanoirEventType.DELETE_DATASET_EVENT, id.toString(), KeycloakUtil.getTokenUserId(), "", ShanoirEvent.SUCCESS, entity.getStudyId()));
+
+	}
+
+	/**
+	 * Call by dataset-details. Also reject from pacs
+	 * @param id dataset id.
+	 * @throws ShanoirException
+	 * @throws SolrServerException
+	 * @throws IOException
+	 * @throws RestServiceException
+	 */
 	@Override
 	@Transactional
 	public void deleteById(final Long id) throws ShanoirException, SolrServerException, IOException, RestServiceException {
 		final Dataset dataset = repository.findById(id)
 				.orElseThrow(() -> new EntityNotFoundException(Dataset.class, id));
 
-		// Do not delete entity if it is the source. If getSourceId() is not null, it means it's a copy
-		List<Dataset> childDs = repository.findBySourceId(id);
-		if (!CollectionUtils.isEmpty(childDs)) {
+		// Do not delete entity if it is the source (or if it has copies). If getSourceId() is not null, it means it's a copy
+		if (!CollectionUtils.isEmpty(dataset.getCopies())) {
 			throw new RestServiceException(
 					new ErrorModel(
 							HttpStatus.UNPROCESSABLE_ENTITY.value(),
 							"This dataset is linked to another dataset that was copied."
 					));
-
 		}
-		// Remove parent processing to avoid errors
-		dataset.setDatasetProcessing(null);
-		processingService.removeDatasetFromAllProcessingInput(id);
-		processingResourceService.deleteByDatasetId(id);
-		propertyService.deleteByDatasetId(id);
-		repository.deleteById(id);
 
-		if (dataset.getSource() != null) {
+		delete(dataset);
+
+		if (CollectionUtils.isEmpty(dataset.getCopies())) {
```

Comment:
not sure, if this "if" is necessary, as you throw an exception in case of not-empty collection getCopies above